README.md
# Rollback Auto Update

* Contributors:      afragen, costdev
* Author:            WP Core Contributors
* License:           MIT
* Requires at least: 6.0
* Requires PHP:      5.6
* Stable tag:        main

Rollback protection from a plugin auto-update whose activation would result in a PHP fatal or error. This is part of the Rollback Update Failure feature project.

## Description
Rollback protection from a plugin auto-update whose activation would result in a PHP fatal or error. This is part of the Rollback Update Failure feature project.

The [Rollback Update Failure](https://wordpress.org/plugins/rollback-update-failure/) plugin must be installed and active.

## Testing
To test you must have a plugin that is capable of being updated **and** whose update contains a PHP fatal error or warning upon activation.

The simplest way to test is to download, install, and activate this plugin and [Git Updater](https://git-updater.com). Then install the test plugin, [Rollback Fatal Plugin](https://github.com/afragen/rollback-fatal-plugin/), via WP-CLI with the following command. 

`wp plugin install https://github.com/afragen/rollback-fatal-plugin/archive/refs/heads/main.zip` 

If you wish to download the zip directly, you will need to unzip, rename the containing folder to remove the `-main`, re-zip, and then install via the Upload Plugin screen.

Once installed you will need to change the version number to `0` and comment out the code. After doing this you can activate the plugin, but you can leave it disabled. Set it to _Enable auto-updates_ in the Plugins page.

If the test plugin is active and if the Rollback works, your site should not show a WSOD after the auto-update as the previous, commented plugin will still be the active plugin. You will still show a plugin update.

An email will be sent to the site admin upon an auto-update failure.


CHANGES.md
[unreleased]
* remove some error logging

#### 0.9.2 / 2022-10-09
* move `send_update_result_email` so fires after `restart_updates`

#### 0.9.1 / 2022-09-05
* more updates to email text

#### 0.9.0 / 2022-09-04
* add `sleep(2)` to possibly prevent a race condition before each plugin and before restarts
* update email text

#### 0.8.0 / 2022-08-24
* large refactor into smaller chunks
* load in static hook
* don't check our own plugin, cause it will always have a PHP fatal for redeclaring the class
* refactor for restarting plugin update process after a fatal
* add function for sending successful update email
* refactor for inclusion into core

#### 0.6.0 / 2022-08-05
* update documentation
* now includes all type of error, exception, shutdown handling
* update some naming
* email site admin on auto-update failure
* add some error logging as the debug.log will fill with errors
* return early if `$hook_extra` not correctly populated
* clean up
* limit scope as much as possible

#### 0.5.0 / 2022-08-02
* converted original Gist to repo


tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file.
 *
 * @package Rollback_Auto_Update
 */

$_tests_dir = getenv( 'WP_TESTS_DIR' );

if ( ! $_tests_dir ) {
	$_tests_dir = rtrim( sys_get_temp_dir(), '/\\' ) . '/wordpress-tests-lib';
}

// Forward custom PHPUnit Polyfills configuration to PHPUnit bootstrap file.
$_phpunit_polyfills_path = getenv( 'WP_TESTS_PHPUNIT_POLYFILLS_PATH' );
if ( false !== $_phpunit_polyfills_path ) {
	define( 'WP_TESTS_PHPUNIT_POLYFILLS_PATH', $_phpunit_polyfills_path );
}

if ( ! file_exists( "{$_tests_dir}/includes/functions.php" ) ) {
	echo "Could not find {$_tests_dir}/includes/functions.php, have you run bin/install-wp-tests.sh ?" . PHP_EOL; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
	exit( 1 );
}

// Give access to tests_add_filter() function.
require_once "{$_tests_dir}/includes/functions.php";

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/plugin.php';
}

tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require "{$_tests_dir}/includes/bootstrap.php";


tests/test-sample.php
<?php
/**
 * Class SampleTest
 *
 * @package Rollback_Auto_Update
 */

/**
 * Sample test case.
 */
class SampleTest extends WP_UnitTestCase {

	/**
	 * A single example test.
	 */
	public function test_sample() {
		// Replace this with some actual testing code.
		$this->assertTrue( true );
	}
}


plugin.php
<?php
/**
 * Rollback Auto Update
 *
 * @author  Andy Fragen, Colin Stewart
 * @license MIT
 * @link    https://github.com/afragen/rollback-auto-update
 * @package rollback-auto-update
 */

/**
 * Rollback an auto-update containing an activation error.
 *
 * @package Rollback_Auto_Update
 *
 * Plugin Name:       Rollback Auto-Update
 * Plugin URI:        https://github.com/afragen/rollback-auto-update
 * Description:       Rollback an auto-update containing an activation error.
 * Version:           0.9.2.1
 * Author:            WP Core Contributors
 * License:           MIT
 * Requires at least: 6.0
 * Requires PHP:      5.6
 * GitHub Plugin URI: https://github.com/afragen/rollback-auto-update
 * Primary Branch:    main
 */

namespace WP_Rollback_Auto_Update;

/*
 * Exit if called directly.
 * PHP version check and exit.
 */
if ( ! defined( 'WPINC' ) ) {
	die;
}

/**
 * Class Init
 */
class Init {

	/**
	 * Let's get started.
	 */
	public function __construct() {
		// Add to wp-admin/includes/admin.php.
		require_once __DIR__ . '/wp-admin/includes/class-rollback-auto-update.php';

		// Add to wp-admin/includes/admin-filters.php.
		add_action( 'init', array( 'WP_Rollback_Auto_Update', 'init' ) );
	}
}

new Init();


rollback-auto-update.php
<?php
/**
 * Rollback Auto Update
 *
 * @author  Andy Fragen, Colin Stewart
 * @license MIT
 * @link    https://github.com/afragen/rollback-auto-update
 * @package rollback-auto-update
 */

/**
 * Rollback an auto-update containing an activation error.
 *
 * @package Rollback_Auto_Update
 *
 * Plugin Name:       Rollback Auto-Update
 * Plugin URI:        https://github.com/afragen/rollback-auto-update
 * Description:       Rollback an auto-update containing an activation error.
 * Version:           0.6.4
 * Author:            WP Core Contributors
 * License:           MIT
 * Requires at least: 5.9
 * Requires PHP:      5.6
 * GitHub Plugin URI: https://github.com/afragen/rollback-auto-update
 * Primary Branch:    main
 */

namespace Fragen;

/*
 * Exit if called directly.
 * PHP version check and exit.
 */
if ( ! defined( 'WPINC' ) ) {
	die;
}

/**
 * Class Rollback_Auto_Update
 */
class Rollback_Auto_Update {

	/**
	 * Stores handler parameters.
	 *
	 * @var array
	 */
	private $handler_args = [];

	/**
	 * Stores `update_plugins` transient.
	 *
	 * @var \stdClass
	 */
	private $current;

	/**
	 * Stores error codes.
	 *
	 * @var int
	 */
	public $error_types = E_ERROR | E_PARSE | E_COMPILE_ERROR | E_USER_ERROR | E_RECOVERABLE_ERROR;

	/**
	 * Constructor, let's get going.
	 */
	public function __construct() {
		add_filter( 'upgrader_install_package_result', [ $this, 'auto_update_check' ], 15, 2 );
	}

	/**
	 * Checks the validity of the updated plugin.
	 *
	 * @param array|WP_Error $result     Result from WP_Upgrader::install_package().
	 * @param array          $hook_extra Extra arguments passed to hooked filters.
	 *
	 * @return array|WP_Error
	 */
	public function auto_update_check( $result, $hook_extra ) {
		if ( is_wp_error( $result ) || ! wp_doing_cron() || ! isset( $hook_extra['plugin'] ) ) {
			return $result;
		}

		$this->current      = get_site_transient( 'update_plugins' );
		$this->handler_args = [
			'handler_error' => '',
			'result'        => $result,
			'hook_extra'    => $hook_extra,
		];
		$processed          = (array) get_site_transient( 'processed_auto_updates' );

		// Register exception and shutdown handlers.
		$this->initialize_handlers();

		// Working parts of `plugin_sandbox_scrape()`.
		wp_register_plugin_realpath( WP_PLUGIN_DIR . '/' . $hook_extra['plugin'] );
		if ( 'rollback-auto-update/rollback-auto-update.php' !== $hook_extra['plugin'] ) {
			include_once WP_PLUGIN_DIR . '/' . $hook_extra['plugin'];
		}

		$processed[] = $hook_extra['plugin'];
		set_site_transient( 'processed_auto_updates', $processed, 60 );

		\error_log( $hook_extra['plugin'] . ' auto updated ' );

		return $result;
	}

	/**
	 * Initializes handlers.
	 */
	private function initialize_handlers() {
		// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_set_error_handler
		set_error_handler( [ $this, 'error_handler' ], ( E_ALL ^ $this->error_types ) );
		set_exception_handler( [ $this, 'exception_handler' ] );
	}

	/**
	 * Handles Errors.
	 */
	public function error_handler() {
		$this->handler_args['handler_error'] = 'Error Caught';
		$this->handler();
	}

	/**
	 * Handles Exceptions.
	 */
	public function exception_handler() {
		$this->handler_args['handler_error'] = 'Exception Caught';
		$this->handler();
	}

	/**
	 * Handles errors by running Rollback.
	 */
	private function handler() {
		// Exit for non-fatal errors.
		$e = error_get_last();
		if ( ! empty( $e ) && $this->error_types !== $e['type'] ) {
			// error_log( 'return on non-fatal error: ' . $this->handler_args['hook_extra']['plugin'] );
			// error_log( 'non-fatal error: ' . \var_export( $e, true ) );
			return;
		}
		set_site_transient( 'rollback_fatal_plugin', [ $this->handler_args['hook_extra']['plugin'] ], 60 );

		$this->cron_rollback();
		$this->log_error_msg( $e );
		$this->restart_updates();
		$this->send_update_result_email();
		return;
	}

	/**
	 * Rolls back during cron.
	 *
	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
	 */
	private function cron_rollback() {
		global $wp_filesystem;

		$temp_backup = [
			'temp_backup' => [
				'dir'  => 'plugins',
				'slug' => dirname( $this->handler_args['hook_extra']['plugin'] ),
				'src'  => $wp_filesystem->wp_plugins_dir(),
			],
		];

		include_once $wp_filesystem->wp_plugins_dir() . 'rollback-update-failure/wp-admin/includes/class-wp-upgrader.php';
		$rollback_updater = new \Rollback_Update_Failure\WP_Upgrader();

		// Set private $temp_restores variable.
		$ref_temp_restores = new \ReflectionProperty( $rollback_updater, 'temp_restores' );
		$ref_temp_restores->setAccessible( true );
		$ref_temp_restores->setValue( $rollback_updater, $temp_backup );

		// Set private $temp_backups variable.
		$ref_temp_backups = new \ReflectionProperty( $rollback_updater, 'temp_backups' );
		$ref_temp_backups->setAccessible( true );
		$ref_temp_backups->setValue( $rollback_updater, $temp_backup );

		// Call Rollback's restore_temp_backup().
		$restore_temp_backup = new \ReflectionMethod( $rollback_updater, 'restore_temp_backup' );
		$restore_temp_backup->invoke( $rollback_updater );

		// Call Rollback's delete_temp_backup().
		$delete_temp_backup = new \ReflectionMethod( $rollback_updater, 'delete_temp_backup' );
		$delete_temp_backup->invoke( $rollback_updater );

		$this->send_fatal_error_email();
	}

	/**
	 * Sends an email to the site administrator when a plugin
	 * new version contains a fatal error.
	 *
	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
	 */
	private function send_fatal_error_email() {
		global $wp_filesystem;

		if ( ! isset( $this->handler_args['hook_extra']['plugin'] ) ) {
			return;
		}

		$plugin_path = $wp_filesystem->wp_plugins_dir() . $this->handler_args['hook_extra']['plugin'];
		$name        = \get_plugin_data( $plugin_path )['Name'];
		$subject     = __( 'A plugin was rolled back to the previously installed version' );
		$body        = sprintf(
			__( 'Howdy!' ) . "\n\n" .
			/* translators: 1: The name of the plugin or theme. 2: Home URL. */
			__( '%1$s was successfully updated on your site, [%2$s], at %3$s.' ) . "\n\n" .
			__( 'However, due to a fatal error, it was reverted to the previously installed version to keep your site running.' ) . ' ' .
			__( 'If a new version is released without fatal errors, it will be installed automatically.' ) . "\n\n",
			$name,
			get_bloginfo( 'name' ),
			home_url()
		);

		$body .= __( 'The WordPress Rollback Team' ) . "\n";

		wp_mail( get_bloginfo( 'admin_email' ), $subject, $body );
	}

	/**
	 * Outputs the handler error to the log file.
	 *
	 * @param array $e Error code.
	 */
	private function log_error_msg( $e ) {
		$error_msg = sprintf(
			'Rollback Auto-Update: %1$s in %2$s, error type %3$s',
			$this->handler_args['handler_error'],
			$this->handler_args['hook_extra']['plugin'],
			empty( $e ) ? 'fatal' : $e['type']
		);
		//phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
		error_log( $error_msg );
	}

	/**
	 * Restart update process for plugins that remain after a fatal.
	 */
	private function restart_updates() {
		$remaining_auto_updates = $this->get_remaining_auto_updates();

		if ( empty( $remaining_auto_updates ) ) {
			return;
		}

		$skin     = new \Automatic_Upgrader_Skin();
		$upgrader = new \Plugin_Upgrader( $skin );
		\error_log( 'Plugin_Upgrader::bulk_upgrade' . "\n" . var_export( $remaining_auto_updates, true ) );
		$upgrader->bulk_upgrade( $remaining_auto_updates );
	}

	/**
	 * Sends an email noting successful and failed updates.
	 */
	private function send_update_result_email() {
		$processed  = array_unique( array_filter( (array) get_site_transient( 'processed_auto_updates' ) ) );
		$fatals     = (array) get_site_transient( 'rollback_fatal_plugin' );
		$successful = $failed = [];

		\error_log( 'send_update_result_email processed: ' . \var_export( $processed, true ) );
		\error_log( 'send_update_result_email fatals: ' . \var_export( $fatals, true ) );
		/*
		 * Using `get_plugin_data()` instead has produced warnings/errors
		 * as the files may not be in place at this time.
		 */
		$plugins = get_plugins();

		foreach ( $this->current->response as $k => $update ) {
			$item = $this->current->response[ $k ];
			$name = $plugins[ $update->plugin ]['Name'];

			/*
			 * This appears to be the only way to get a plugin's older version
			 * at this stage of an auto-update when not implementing this
			 * feature directly in Core.
			 */
			$current_version = $this->current->checked[ $update->plugin ];

			/*
			 * The `current_version` property does not exist yet. Add it.
			 *
			 * `$this->current->response[ $k ]` is an instance of `stdClass`,
			 * so this should not fall victim to PHP 8.2's deprecation of
			 * dynamic properties.
			 */
			$item->current_version = $current_version;

			$plugin_result = (object) [
				'name' => $name,
				'item' => $item,
			];

			if ( in_array( $update->plugin, $processed, true ) ) {
				$successful['plugin'][] = $plugin_result;
				continue;
			}

			if ( in_array( $update->plugin, $fatals, true ) ) {
				$failed['plugin'][] = $plugin_result;
			}
		}

		$automatic_upgrader      = new \WP_Automatic_Updater();
		$send_plugin_theme_email = new \ReflectionMethod( $automatic_upgrader, 'send_plugin_theme_email' );
		$send_plugin_theme_email->setAccessible( true );
		$send_plugin_theme_email->invoke( $automatic_upgrader, 'mixed', $successful, $failed );
	}

	/**
	 * Get array of non-fataling auto-updates remaining.
	 *
	 * @return array
	 */
	private function get_remaining_auto_updates() {
		if ( empty( $this->handler_args ) ) {
			return [];
		}
		$processed = (array) get_site_transient( 'processed_auto_updates' );
		$fatals    = (array) get_site_transient( 'rollback_fatal_plugin' );

		// Get array of plugins set for auto-updating.
		$auto_updates    = (array) get_site_option( 'auto_update_plugins', [] );
		$current_plugins = array_keys( $this->current->response );

		// Get all auto-updating plugins that have updates available.
		$current_auto_updates = array_intersect( $auto_updates, $current_plugins );
		error_log( 'current_auto_updates ' . var_export( $current_auto_updates, true ) );
		error_log( 'fatals ' . var_export( $fatals, true ) );

		// Get array of non-fatal auto-updates remaining.
		$remaining_auto_updates = array_diff( $current_auto_updates, $processed, $fatals );

		$processed = array_unique( array_merge( $processed, $remaining_auto_updates ) );
		\set_site_transient( 'processed_auto_updates', $processed, 60 );
		error_log( 'remaining_auto_updates ' . var_export( $remaining_auto_updates, true ) );

		return $remaining_auto_updates;
	}
}

new Rollback_Auto_Update();


auto-update-fatal-error-rollback.php
<?php
/**
 * Rollback an auto-update containing an activation error.
 *
 * @package Rollback_Auto_Update
 *
 * Plugin Name:       Rollback Auto-Update
 * Plugin URI:        https://github.com/afragen/auto-update-fatal-error-rollback
 * Description:       Rollback an auto-update containing an activation error.
 * Version:           0.5.5
 * Author:            WP Core Contributors
 * License:           MIT
 * Requires at least: 5.9
 * Requires PHP:      5.6
 * GitHub Plugin URI: https://github.com/afragen/auto-update-fatal-error-rollback
 * Primary Branch:    main
 */

namespace Fragen;

add_filter(
	'upgrader_install_package_result',
	[ new Rollback_Auto_Update(), 'auto_update_check' ],
	10,
	2
);

/**
 * Class Auto_Update_Failure_Check
 */
class Rollback_Auto_Update {

	/**
	 * Check validity of updated plugin.
	 *
	 * @param array|WP_Error $result     Result from WP_Upgrader::install_package().
	 * @param array          $hook_extra Extra arguments passed to hooked filters.
	 *
	 * @return array|WP_Error
	 */
	public function auto_update_check( $result, $hook_extra ) {
		if ( ! is_wp_error( $result ) && wp_doing_cron() ) {
			$plugin = $hook_extra['plugin'];

			// Register exception and shutdown handlers.
			$handler_args = [
				'error'      => 'Shutdown Caught',
				'result'     => $result,
				'hook_extra' => $hook_extra,
			];
			$lambda_error = function( $error ) use ( $handler_args ) {
				$handler_args['error'] = 'Error Caught';
				$this->handler( $handler_args );
			};
			// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_set_error_handler
			set_error_handler( $lambda_error );
			$lambda_exception = function( $exception ) use ( $handler_args ) {
				$handler_args['error'] = 'Exception Caught';
				$this->handler( $handler_args );
			};
			set_exception_handler( $lambda_exception );
			register_shutdown_function( [ $this, 'handler' ], $handler_args );

			// working parts of `plugin_sandbox_scrape()`.
			wp_register_plugin_realpath( WP_PLUGIN_DIR . '/' . $plugin );
			include WP_PLUGIN_DIR . '/' . $plugin;
		}

		return $result;
	}

	/**
	 * Rollback during cron.
	 *
	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
	 *
	 * @param array|WP_Error $result     Result from WP_Upgrader::install_package().
	 * @param array          $hook_extra Extra arguments passed to hooked filters.
	 *
	 * @return array|WP_Error
	 */
	public function cron_rollback( $result, $hook_extra ) {
		global $wp_filesystem;

		if ( ! isset( $hook_extra['plugin'] ) ) {
			return $result;
		}
		$result      = new \WP_Error( 'unexpected_output', __( 'The plugin generated unexpected output.' ) );
		$plugin      = $hook_extra['plugin'];
		$temp_backup = [
			'temp_backup' => [
				'dir'  => 'plugins',
				'slug' => dirname( $plugin ),
				'src'  => $wp_filesystem->wp_plugins_dir(),
			],
		];
		$hook_extra  = array_merge( $hook_extra, $temp_backup );
		include_once $wp_filesystem->wp_plugins_dir() . 'rollback-update-failure/wp-admin/includes/class-wp-upgrader.php';
		$rollback_updater = new \Rollback_Update_Failure\WP_Upgrader();

		// Set private $temp_restores variable.
		$ref_temp_restores = new \ReflectionProperty( $rollback_updater, 'temp_restores' );
		$ref_temp_restores->setAccessible( true );
		$ref_temp_restores->setValue( $rollback_updater, $temp_backup );

		// Set private $temp_backups variable.
		$ref_temp_backups = new \ReflectionProperty( $rollback_updater, 'temp_backups' );
		$ref_temp_backups->setAccessible( true );
		$ref_temp_backups->setValue( $rollback_updater, $temp_backup );

		// Call Rollback's restore_temp_backup().
		$restore_temp_backup = new \ReflectionMethod( $rollback_updater, 'restore_temp_backup' );
		$restore_temp_backup->invoke( $rollback_updater );

		// Call Rollback's delete_temp_backup().
		$delete_temp_backup = new \ReflectionMethod( $rollback_updater, 'delete_temp_backup' );
		$delete_temp_backup->invoke( $rollback_updater );

		return $result;
	}

	/**
	 * Handle errors by running Rollback.
	 *
	 * @param array $args Array of args.
	 *
	 * @return void
	 */
	public function handler( $args ) {
		\error_log( $args['error'] );
		$this->cron_rollback( $args['result'], $args['hook_extra'] );
	}
}


wp-admin/includes/class-rollback-auto-update.php
<?php
/**
 * WordPress Plugin Administration API: WP_Rollback_Auto_Update class
 *
 * @package WordPress
 * @subpackage Administration
 * @since 6.2.0
 */

/**
 * Core class for rolling back auto-update plugin failures.
 */
class WP_Rollback_Auto_Update {

	/**
	 * Stores handler parameters.
	 *
	 * @var array
	 */
	private $handler_args = array();

	/**
	 * Stores successfully updated plugins.
	 *
	 * @var array
	 */
	private $processed = array();

	/**
	 * Stores fataling plugins.
	 *
	 * @var array
	 */
	private $fatals = array();

	/**
	 * Stores `update_plugins` transient.
	 *
	 * @var \stdClass
	 */
	private static $current;

	/**
	 * Stores boolean for no error from check_plugin_for_errors().
	 *
	 * @var bool
	 */
	private $no_error = false;

	/**
	 * Stores error codes.
	 *
	 * @var int
	 */
	public $error_types = E_ERROR | E_PARSE | E_COMPILE_ERROR | E_USER_ERROR | E_RECOVERABLE_ERROR;

	/**
	 * Static function to get started from hook.
	 *
	 * @return void
	 */
	public static function init() {
		( new WP_Rollback_Auto_Update() )->load_hooks();
	}

	/**
	 * Load hook to start.
	 *
	 * @return void
	 */
	public function load_hooks() {
		add_filter( 'upgrader_install_package_result', array( $this, 'auto_update_check' ), 15, 2 );
	}

	/**
	 * Checks the validity of the updated plugin.
	 *
	 * @param array|WP_Error $result     Result from WP_Upgrader::install_package().
	 * @param array          $hook_extra Extra arguments passed to hooked filters.
	 *
	 * @return array|WP_Error
	 */
	public function auto_update_check( $result, $hook_extra ) {
		if ( is_wp_error( $result ) || ! wp_doing_cron() || ! isset( $hook_extra['plugin'] ) ) {
			return $result;
		}

		// Checking our own plugin will cause a PHP Fatal redeclaration error.
		if ( 'rollback-auto-update/plugin.php' === $hook_extra['plugin'] ) {
			return $result;
		}
		error_log( $hook_extra['plugin'] . ' processing...' );

		// This possibly helps to avoid a potential race condition on servers that may start to
		// process the next plugin for auto-updating before the handler can pick up an error from
		// the previously processed plugin.
		sleep( 2 );

		$this->no_error     = false;
		static::$current    = get_site_transient( 'update_plugins' );
		$this->handler_args = array(
			'handler_error' => '',
			'result'        => $result,
			'hook_extra'    => $hook_extra,
		);

		// Register exception and shutdown handlers.
		$this->initialize_handlers();

		$this->processed[] = $hook_extra['plugin'];

		if ( is_plugin_inactive( $hook_extra['plugin'] ) ) {
			// Working parts of plugin_sandbox_scrape().
			wp_register_plugin_realpath( WP_PLUGIN_DIR . '/' . $hook_extra['plugin'] );
			include WP_PLUGIN_DIR . '/' . $hook_extra['plugin'];
		}

		// Needs to run for both active and inactive plugins. Don't ask why, just accept it.
		$this->check_plugin_for_errors( $hook_extra['plugin'] );
		error_log( $hook_extra['plugin'] . ' auto updated.' );

		return $result;
	}

	/**
	 * Checks a new plugin version for errors.
	 *
	 * If an error is found, the previously installed version will be reinstalled
	 * and an email will be sent to the site administrator.
	 *
	 * @param string $plugin The plugin to check.
	 *
	 * @throws \Exception If errors are present.
	 *
	 * @return void
	 */
	private function check_plugin_for_errors( $plugin ) {
		global $wp_filesystem;

		if ( file_exists( ABSPATH . '.maintenance' ) ) {
			$wp_filesystem->delete( ABSPATH . '.maintenance' );
		}

		$nonce    = wp_create_nonce( 'plugin-activation-error_' . $plugin );
		$response = wp_remote_get(
			add_query_arg(
				array(
					'action'   => 'error_scrape',
					'plugin'   => $plugin,
					'_wpnonce' => $nonce,
				),
				admin_url( 'plugins.php' )
			),
			array( 'timeout' => 60 )
		);

		if ( is_wp_error( $response ) ) {
			// If it isn't possible to run the check, assume an error.
			throw new \Exception( $response->get_error_message() );
		}

		$code           = wp_remote_retrieve_response_code( $response );
		$body           = wp_remote_retrieve_body( $response );
		$this->no_error = 200 === $code;

		// error_log( 'check_for_plugin_errors code: ' . var_export( $code, true ) );
		if ( str_contains( $body, 'wp-die-message' ) || 200 !== $code ) {
			$error = new \WP_Error(
				'new_version_error',
				sprintf(
					/* translators: %s: The name of the plugin. */
					__( 'The new version of %s contains an error' ),
					\get_plugin_data( WP_PLUGIN_DIR . '/' . $plugin )['Name']
				)
			);
			throw new \Exception( $error->get_error_message() );
		}
	}

	/**
	 * Initializes handlers.
	 */
	private function initialize_handlers() {
		// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_set_error_handler
		set_error_handler( array( $this, 'error_handler' ), ( E_ALL ^ $this->error_types ) );
		set_exception_handler( array( $this, 'exception_handler' ) );
	}

	/**
	 * Handles Errors.
	 */
	public function error_handler() {
		$this->handler_args['handler_error'] = 'Error Caught';
		// error_log( 'error handler: ' . var_export( $this->handler_args['hook_extra']['plugin'], true ) );

		$this->handler( $this->non_fatal_errors() );
	}

	/**
	 * Handles Exceptions.
	 */
	public function exception_handler() {
		$this->handler_args['handler_error'] = 'Exception Caught';
		// error_log( 'exception handler: ' . var_export( $this->handler_args['hook_extra']['plugin'], true ) );
		$this->handler( false );
	}

	/**
	 * Handles errors by running Rollback.
	 *
	 * @param bool $skip If false, assume fatal and process, default false.
	 */
	private function handler( $skip = false ) {
		if ( $skip ) {
			// error_log( 'exit handler skip: ' . var_export( $skip, true ) );
			return;
		}
		$this->fatals[] = $this->handler_args['hook_extra']['plugin'];

		$this->cron_rollback();
		$this->log_error_msg( error_get_last() );

		// Let's sleep for a couple of seconds here.
		// After the error handler and before restarting updates.
		sleep( 2 );

		$this->restart_updates();
		$this->send_update_result_email();
	}

	/**
	 * Skip (exit handler() early) for non-fatal errors or non-errors.
	 *
	 * @return bool
	 */
	private function non_fatal_errors() {
		$e                = error_get_last();
		$non_fatal_errors = ( ! empty( $e ) && $this->error_types !== $e['type'] );
		$skip             = is_plugin_active( $this->handler_args['hook_extra']['plugin'] ) || $this->no_error;
		$skip             = $skip ? $skip : $non_fatal_errors;

		return $skip;
	}

	/**
	 * Rolls back during cron.
	 *
	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
	 */
	private function cron_rollback() {
		global $wp_filesystem;

		$temp_backup = array(
			'temp_backup' => array(
				'dir'  => 'plugins',
				'slug' => dirname( $this->handler_args['hook_extra']['plugin'] ),
				'src'  => $wp_filesystem->wp_plugins_dir(),
			),
		);

		include_once $wp_filesystem->wp_plugins_dir() . 'rollback-update-failure/wp-admin/includes/class-wp-upgrader.php';
		$rollback_updater = new \Rollback_Update_Failure\WP_Upgrader();

		// Set private $temp_restores variable.
		$ref_temp_restores = new \ReflectionProperty( $rollback_updater, 'temp_restores' );
		$ref_temp_restores->setAccessible( true );
		$ref_temp_restores->setValue( $rollback_updater, $temp_backup );

		// Set private $temp_backups variable.
		$ref_temp_backups = new \ReflectionProperty( $rollback_updater, 'temp_backups' );
		$ref_temp_backups->setAccessible( true );
		$ref_temp_backups->setValue( $rollback_updater, $temp_backup );

		// Call Rollback's restore_temp_backup().
		$restore_temp_backup = new \ReflectionMethod( $rollback_updater, 'restore_temp_backup' );
		$restore_temp_backup->invoke( $rollback_updater );

		// Call Rollback's delete_temp_backup().
		$delete_temp_backup = new \ReflectionMethod( $rollback_updater, 'delete_temp_backup' );
		$delete_temp_backup->invoke( $rollback_updater );

		error_log( $this->handler_args['hook_extra']['plugin'] . ' rolled back' );
	}

	/**
	 * Outputs the handler error to the log file.
	 *
	 * @param array $e Last error.
	 */
	private function log_error_msg( $e ) {
		$error_msg = sprintf(
			'Rollback Auto-Update: %1$s in %2$s, error %3$s',
			$this->handler_args['handler_error'],
			$this->handler_args['hook_extra']['plugin'],
			empty( $e ) ? '?ParseError?' : var_export( $e, true )
		);
		//phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
		error_log( $error_msg );
	}

	/**
	 * Restart update process for plugins that remain after a fatal.
	 */
	private function restart_updates() {
		$remaining_auto_updates = $this->get_remaining_auto_updates();

		if ( empty( $remaining_auto_updates ) ) {
			return;
		}

		// error_log( 'restart Plugin_Upgrader::bulk_upgrade ' . var_export( $remaining_auto_updates, true ) );
		$skin     = new \Automatic_Upgrader_Skin();
		$upgrader = new \Plugin_Upgrader( $skin );
		$upgrader->bulk_upgrade( $remaining_auto_updates );
	}

	/**
	 * Get array of non-fataling auto-updates remaining.
	 *
	 * @return array
	 */
	private function get_remaining_auto_updates() {
		if ( empty( $this->handler_args ) ) {
			return array();
		}

		// Get array of plugins set for auto-updating.
		$auto_updates    = (array) get_site_option( 'auto_update_plugins', array() );
		$current_plugins = array_keys( static::$current->response );

		// Get all auto-updating plugins that have updates available.
		$current_auto_updates = array_intersect( $auto_updates, $current_plugins );

		// Get array of non-fatal auto-updates remaining.
		$remaining_auto_updates = array_diff( $current_auto_updates, $this->processed, $this->fatals );

		$this->processed = array_unique( array_merge( $this->processed, $remaining_auto_updates ) );

		// error_log( 'fatals ' . var_export( array_unique( $this->fatals ), true ) );
		// error_log( 'current auto updates: ' . var_export( $current_auto_updates, true ) );
		// error_log( 'remaining auto updates ' . var_export( $remaining_auto_updates, true ) );
		// error_log( 'processed ' . var_export( $this->processed, true ) );
		return $remaining_auto_updates;
	}

	/**
	 * Sends an email noting successful and failed updates.
	 */
	private function send_update_result_email() {
		add_filter( 'auto_plugin_theme_update_email', array( $this, 'props' ), 10, 4 );
		$successful = array();
		$failed     = array();

		/*
		 * Using `get_plugin_data()` instead has produced warnings/errors
		 * as the files may not be in place at this time.
		 */
		$plugins = get_plugins();

		foreach ( static::$current->response as $k => $update ) {
			$item = static::$current->response[ $k ];
			$name = $plugins[ $update->plugin ]['Name'];

			/*
			 * This appears to be the only way to get a plugin's older version
			 * at this stage of an auto-update when not implementing this
			 * feature directly in Core.
			 */
			$current_version = static::$current->checked[ $update->plugin ];

			/*
			 * The `current_version` property does not exist yet. Add it.
			 *
			 * `static::$current->response[ $k ]` is an instance of `stdClass`,
			 * so this should not fall victim to PHP 8.2's deprecation of
			 * dynamic properties.
			 */
			$item->current_version = $current_version;

			$plugin_result = (object) array(
				'name' => $name,
				'item' => $item,
			);

			$success = array_diff( $this->processed, $this->fatals );

			if ( in_array( $update->plugin, $success, true ) ) {
				$successful['plugin'][] = $plugin_result;
				continue;
			}

			if ( in_array( $update->plugin, $this->fatals, true ) ) {
				$failed['plugin'][] = $plugin_result;
			}
		}

		$automatic_upgrader      = new \WP_Automatic_Updater();
		$send_plugin_theme_email = new \ReflectionMethod( $automatic_upgrader, 'send_plugin_theme_email' );
		$send_plugin_theme_email->setAccessible( true );
		$send_plugin_theme_email->invoke( $automatic_upgrader, 'mixed', $successful, $failed );
	}

	/**
	 * Credit where credit is due.
	 *
	 * @param array  $email {
	 *     Array of email arguments that will be passed to wp_mail().
	 *
	 *     @type string $to      The email recipient. An array of emails
	 *                           can be returned, as handled by wp_mail().
	 *     @type string $subject The email's subject.
	 *     @type string $body    The email message body.
	 *     @type string $headers Any email headers, defaults to no headers.
	 * }
	 * @param string $type               The type of email being sent. Can be one of 'success', 'fail', 'mixed'.
	 * @param array  $successful_updates A list of updates that succeeded.
	 * @param array  $failed_updates     A list of updates that failed.
	 *
	 * @return array
	 */
	public function props( $email, $type, $successful_updates, $failed_updates ) {
		if ( empty( $failed_updates ) ) {
			return $email;
		}
		$body   = explode( "\n", $email['body'] );
		$failed = __( 'These plugins failed to update or may have been rolled back due to detection of a fatal error:' );
		array_splice( $body, 6, 1, $failed );
		$props = __( 'The WordPress Rollbackenberg Team' );
		array_splice( $body, -1, 1, $props );
		$body          = implode( "\n", $body );
		$email['body'] = $body;

		return $email;
	}
}
